internal/servers/controller/internal/marshaler/attributes.go

```diff
@@ -0,0 +1,136 @@
+package marshaler
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/hashicorp/boundary/sdk/pbs/controller/protooptions"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+func init() {
+	globalAttributeKeys = attributeKeys{
+		m: make(map[string]keyMap),
+	}
+
+	protoregistry.GlobalTypes.RangeMessages(func(m protoreflect.MessageType) bool {
+		d := m.Descriptor()
+		if err := globalAttributeKeys.register(d); err != nil {
+			panic(err)
+		}
+		return true
+	})
+}
+
+type keyMap map[string]string
+
+type attributeKeys struct {
+	sync.RWMutex
+	m map[string]keyMap
+}
+
+var globalAttributeKeys attributeKeys
+
+// register examines the given protobuf MessageDescriptor for fields that have
+// the Subtype protobuf extension. It uses these to build a keyMap of subtype
+// strings to the protobuf JSON keys. If the message does not have any subtypes
+// it will not be registered, but no error is returned, allowing this to be
+// called on any protobuf message. However, if a message has subtpyes but does
+// not provide a field with a subtype of "default" an error is returned.
+func (ak attributeKeys) register(d protoreflect.MessageDescriptor) error {
+	ak.Lock()
+	defer ak.Unlock()
```

Comment:
All of the writes *should* happen in this init and happen serially. Then it should just be reads, so even though those reads will happen in separate goroutines it should technically be safe since it is just reads. However, in the past we have added extra precautions particularly around package level maps since there is nothing to enforce that all of the writes happen in init. The lock ensures that it is concurrent-safe and since it is a RWMutex it should have little impact for all the reads.